### ISSUE TITLE
Ad guard for type intersection overrides during declaration

### ISSUE THREAD

**Issue by ramik91** [microsoft/TypeScript#34643]
## Search Terms

- Intersection types
- Guards

## Suggestion

In the following code snippet intersection type `PropsType` won't give an error during the type declaration, but TSC will complain if one try to use it when defining the variable. In case of using an interface instead of intersection type TS will complain during the declaration which seems much more correct to me.

TSC should throw an error when there is an attempt to redefine existing property of one of the existing types using type intersection technique.

```typescript
type BaseProps = {
    foo: boolean;
};

type PropsType = BaseProps & {
    name: string;
    // should be complaining at this point
    foo: number;
};

const props: PropsType = {
    name: 'John',
   // but complains here instead
    foo: 5,
}

// complains correctly about redefining foo as a number
interface IProps extends Base {
    name: string;
    foo: number;
}
```

## Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [ ] This wouldn't change the runtime behavior of existing JavaScript code
* [ ] This could be implemented without emitting different JS based on the types of the expressions
* [ ] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [ ] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by RyanCavanaugh** [comment]
In general this isn't possible, and arguably isn't even desirable.

These intersections produce `never` and it's very common for them to correctly evaporate out of unions in higher-order scenarios in ways that are useful.

Regardless, TypeScript doesn't (and *can't*!) aggressively probe every property type implied by a declaration, so there's no point in time at which TS is guaranteed to even manifest these properties to produce an error in the first place.
